consensus/polybft/polybft.go

```diff
@@ -0,0 +1,662 @@
+
+		for {
+			select {
+			case ev := <-eventCh:
+				currentBlockNum := p.blockchain.CurrentHeader().Number
+				if ev.Source == "syncer" {
+					if ev.NewChain[0].Number < currentBlockNum {
+						continue
+					}
+				}
+
+				if p.isSynced() {
+					syncerBlockCh <- currentBlockNum
+				}
+
+			case <-p.closeCh:
+				return
+			}
+		}
+	}()
+
+	defer newBlockSub.Close()
+
+SYNC:
+	if !p.isSynced() {
+		<-syncerBlockCh
+	}
+
+	lastBlock := p.blockchain.CurrentHeader()
+	p.logger.Info("startPbftProcess",
+		"header hash", lastBlock.Hash,
+		"computed hash", lastBlock.Hash,
+		"header number", lastBlock.Number)
+
+	currentValidators, err := p.GetValidators(lastBlock.Number, nil)
+	if err != nil {
+		p.logger.Error("failed to query current validator set", "block number", lastBlock.Number, "error", err)
+	}
+
+	p.runtime.setIsActiveValidator(currentValidators.ContainsNodeID(p.key.NodeID()))
+
+	if !p.runtime.isActiveValidator() {
+		// inactive validator is not part of the consensus protocol and it should just perform syncing
+		goto SYNC
+	}
+
+	// we have to start the bridge snapshot when we have finished syncing
+	if err := p.runtime.restartEpoch(lastBlock); err != nil {
+		p.logger.Error("failed to restart epoch", "error", err)
+
+		goto SYNC
+	}
+
+	for {
+		if err := p.runCycle(); err != nil {
+			if errors.Is(err, errNotAValidator) {
+				p.logger.Info("Node is no longer in validator set")
+			} else {
+				p.logger.Error("an error occurred while running a state machine cycle.", "error", err)
+			}
+
+			goto SYNC
+		}
+
+		switch p.pbft.GetState() {
+		case pbft.SyncState:
+			// we need to go back to sync
+			goto SYNC
+		case pbft.DoneState:
+			// everything worked, move to the next iteration
+		default:
+			// stopped
+			return
+		}
+	}
+}
+
+// isSynced return true if the current header from the local storage corresponds to the highest block of syncer
+func (p *Polybft) isSynced() bool {
+	// TODO: Check could we change following condition to this:
+	// p.syncer.GetSyncProgression().CurrentBlock >= p.syncer.GetSyncProgression().HighestBlock
+	syncProgression := p.syncer.GetSyncProgression()
+
+	return syncProgression == nil ||
+		p.blockchain.CurrentHeader().Number >= syncProgression.HighestBlock
+}
+
+// runCycle runs a single cycle of the state machine and indicates if node should exit the consensus or keep on running
+func (p *Polybft) runCycle() error {
+	ff, err := p.runtime.FSM()
+	if err != nil {
+		return err
+	}
+
+	if err = p.pbft.SetBackend(ff); err != nil {
+		return err
+	}
+
+	// this cancel is not sexy
+	ctx, cancelFn := context.WithCancel(context.Background())
+
+	go func() {
+		<-p.closeCh
+		cancelFn()
+	}()
+
+	p.pbft.Run(ctx)
+
+	return nil
+}
+
+func (p *Polybft) waitForNPeers() bool {
+	for {
+		select {
+		case <-p.closeCh:
+			return false
+		case <-time.After(2 * time.Second):
+		}
+
+		numPeers := len(p.config.Network.Peers())
+		if numPeers >= minSyncPeers {
+			break
+		}
```

Comment:
This was made for testing so that we could skip the first Round Change when the network was initializing.

What is the use case of running a single-node network of an instant finally consensus protocol?